---
title: "5a. Species vs Proteolytic Metabolites"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load data}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(microbiome)
library(ggcorrplot)
library(scales)

# Taxonomy Data - CLR transformed
ps = readRDS("../Reference Files/taxonomy_phyloseq.rds") %>% microbiome::transform('clr')
# Serum CRP data (ELISA)
crp = read.csv('../Reference Files/serum_crp.csv')
# Significant microbially-derived metabolites from previous paper (serum metabolomics)
# p-cresol and phenylacetylglutamine
met = read.csv('../Reference Files/normalized_proteolytic_metabolites.csv')

# Load functions that allow high-throughput Spearman & GLM analysis
source('999. Cross-Sectional Correlation Functions.R')
```

```{r Prep dataset}
# Load sig taxa
temp2 = readxl::read_xlsx('../Results/Supplementary Data/Differential Abundance/Taxonomy_Multivar_Summary.xlsx',sheet='Species') %>%
  filter(sig_in_at_least_n==1, Variable=='StatusPD')

# Make a vector of the species to test
# Include F. prausnitzii because of its strong association with 
#  Ctrl-associated pathways (Table 2)
psig = readxl::read_xlsx('../Results/Table 1 - Stratified Hits.xlsx') %>%
  filter(`Total (%)`>=10) %>% # at least 10% of the functional hits reproduced
  mutate(Species = sapply(Species,function(x) paste0('s__',x) %>% str_replace_all(' ','_'))) %>% 
  pull(Species)
sig = c(temp2[['Species']],psig) %>% unique

# Select species and melt phyloseq object
other_levels = colnames(ps@tax_table)[colnames(ps@tax_table) != 'Species']
psm = ps %>% subset_taxa(Species %in% sig) %>% psmelt() %>%
  select(Species,everything()) %>% select(-any_of(other_levels),-OTU) %>%
  pivot_wider(names_from = Species, values_from = Abundance)

# Combine with CRP and metabolomic data
psm = psm %>% left_join(crp) %>% left_join(met)

# Clean up a little
rm(crp, met, temp2)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Univariate Analysis

```{r}
# Filter out NA values
temp = psm %>% filter(!is.na(pcresol)); temp2 = psm %>% filter(!is.na(Log2CRP))
spearman.lax = correlate_met(temp, 'pcresol',grp='All') %>% 
  rbind(correlate_met(temp, 'phenylacetylglutamine',grp='All')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl')) %>% 
  rbind(correlate_met(temp2,'Log2CRP',grp='All')) %>% 
  rbind(correlate_met(temp2 %>% filter(Status=='PD'),'Log2CRP',grp='PD')) %>% 
  rbind(correlate_met(temp2 %>% filter(Status=='Ctrl'),'Log2CRP',grp='Ctrl'))
```

```{r SET ORDER FOR ALL PLOTS}
# Arrange metabolites by p-cresol estimates, group 'all' ~~~~~~~~~~~~~
species_order = spearman.lax %>% filter(group=='All',variable=='pcresol') %>% 
  arrange(rho)
species_order$taxon = str_remove_all(species_order$taxon,'s__') %>% str_replace_all('_',' ')
```

```{r Figure}
# Edit levels for plotting
spearman.lax[spearman.lax=='pcresol'] = 'p-Cresol'
spearman.lax[spearman.lax=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
spearman.lax[spearman.lax=='Log2CRP'] = 'Log 2 CRP'
spearman.lax$taxon = str_remove_all(spearman.lax$taxon,'s__') %>% str_replace_all('_',' ')

# Plotting function - labels the plot automatically
# vars = the variables you want to include in the plot
prep_plot = function(vars){
  p = spearman.lax %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(variable %in% vars) %>% 
    mutate(variable = ifelse(variable=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',variable)) %>% 
    ggplot(aes(group,taxon,col = rho)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 10, col = 'white',nudge_y = -0.25) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('variable')
  return(p)
}

spearman.lax$taxon = factor(spearman.lax$taxon, levels = species_order$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))

# Plot CRP
p2 = prep_plot('Log 2 CRP')

ggarrange(plotlist = list(p1,p2),ncol = 2, nrow=1,widths=c(4,3))

ggsave('../Results/Supplementary Figures/Fig. 3ab - Species vs Metabolites - Univar.png',height=5.5,width=14)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Multivariate Analysis 

```{r}
# Filter out NA values
temp = psm %>% filter(!is.na(pcresol)); temp2 = psm %>% filter(!is.na(Log2CRP))
glm.lax = glm_met(temp,'pcresol',grp='All') %>% 
  rbind(glm_met(temp,'phenylacetylglutamine',grp='All')) %>% 
  rbind(glm_met(temp %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(glm_met(temp %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(glm_met(temp %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(glm_met(temp %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl')) %>% 
  rbind(glm_met(temp2,'Log2CRP',grp='All')) %>% 
  rbind(glm_met(temp2 %>% filter(Status=='PD'),'Log2CRP',grp='PD')) %>% 
  rbind(glm_met(temp2 %>% filter(Status=='Ctrl'),'Log2CRP',grp='Ctrl'))

# Extract just the relevant rows for plotting
glm_species = glm.lax %>% filter(str_detect(Explanatory,'Microbe'))
```

```{r Figure}
# Edit levels for plotting
glm_species[glm_species=='pcresol'] = 'p-Cresol'
glm_species[glm_species=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
glm_species[glm_species=='Log2CRP'] = 'Log 2 CRP'
glm_species$Taxon = str_remove_all(glm_species$Taxon,'s__') %>% str_replace_all('_',' ')

# Plotting function - labels the plot automatically
# vars = the variables you want to include in the plot
prep_plot = function(vars){
  p = glm_species %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(Response %in% vars) %>% 
    mutate(Response = ifelse(Response=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',Response)) %>% 
    ggplot(aes(group,Taxon,col = Estimate)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 10, col = 'white',nudge_y = -0.25) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('Response')
  return(p)
}

glm_species$Taxon = factor(glm_species$Taxon, levels = species_order$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))

# Plot CRP
p2 = prep_plot('Log 2 CRP')

ggarrange(plotlist = list(p1,p2),ncol = 2, nrow=1,widths=c(4,3))
ggsave('../Results/3. Metabolite Associations/Fig. 3ab - Species vs Metabolites - Multivar.png',height=5.5,width=14)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Save Stats}
writexl::write_xlsx(list('Univar' = spearman.lax,
                         'Multivar' = glm.lax, 
                         'Multivar Species' = glm.lax %>% filter(str_detect(Explanatory,'Microbe'))),
                    '../Results/Supplementary Data/Metabolite Associations/Species vs Metabolites.xlsx')
```