---
title: "5. Species with vs Metabolites and Disease Progression"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load data}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(microbiome)
library(ggcorrplot)
library(lme4)
library(lmerTest)
library(scales)

source('Supp Scripts/5.1. Load Datasets.R')
source('Supp Scripts/5.3. LMER Functions.R')
source('Supp Scripts/5.4. Cross-Sectional Functions.R')
```

```{r Metabolite (amino acid) data}
# Other data
aa = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/Reference Files/Metadata/aa_metabolites.csv') %>% 
  rename(ID = Record.ID) %>% mutate(ID = as.character(ID))
id = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/Reference Files/ID_key.csv')

met = aa %>% rename(record_id = ID) %>% 
  left_join(id)
met$record_id[!is.na(met$correct_id)] = met$correct_id[!is.na(met$correct_id)]
met = met %>% rename(Sample = record_id) %>% select(-correct_id)
met$Sample = paste('X_',met$Sample,sep='')
rm(aa,id)

# Join taxonomy data
met = met %>% left_join(temp) %>% 
  # Remove anyone without taxonomy data
  filter(!is.na(s__Ruthenibacterium_lactatiformans))
```

```{r Longitudinal Data}
# Load dataset, choose relevant metadata
long0 = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/Reference Files/Metadata/shayanandkiradata_nov22.csv') %>% 
  select(Record.ID,redcap_event_name,Disease.duration,bdi_total,contains('mds'),td.pigd,fss_total,months_last_visit) %>% 
  rename(Sample = Record.ID) %>% 
  # Add an X to sample names to match the other datasets
  mutate(Sample = paste0('X_',Sample)) %>% 
  arrange(months_last_visit) %>% 
  group_by(Sample) %>% 
  # Set the baseline as 0 months
  mutate(months_last_visit = ifelse(redcap_event_name=='visit_1_arm_1',0,months_last_visit)) %>%
  # Select only those with longitudinal data available
  add_count() %>% filter(n>1) %>% select(-n) %>%
  ungroup

# Select microbiome-relevant metadata, join to longitudinal dataset
micro = psm %>% select(Sample,Status,Sex,bristol,ent_boolean,depth,OTU,Abundance,Kingdom:Species) %>% unique
long = long0 %>% left_join(micro)
```

# *We now have two datasets: met (metabolomics) and long (longitudinal analysis)*

Both have clr-transformed species of interest attached.

'sig' can be used as a vector of all significant species.

# PART 1: CROSS-SECTIONAL SPECIES VS METABOLITES~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Ctrl and PD tested separately and together.

met used for amino acids, crp used for crp

```{r Spearman}
met2 = met %>% mutate(bristol_grp = ifelse(is.na(bristol),NA,ifelse(bristol<3,'Firm',ifelse(bristol<5,'Normal','Loose')))) %>% 
  mutate_at(vars('pcresol','phenylacetylglutamine','Log2CRP'),function(x)rescale(x))
met2$bristol_grp = factor(met2$bristol_grp, levels=c('Loose','Normal','Firm'))

met2 = met2 %>% pivot_longer(cols = c('pcresol','phenylacetylglutamine','Log2CRP'),
                             names_to = 'Variable', values_to = 'value') %>% 
  filter(!is.na(value))

met2[met2=='pcresol'] = 'p-Cresol'
met2[met2=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
met2[met2=='Log2CRP'] = 'Log 2 CRP'

table(met2$Status,met2$bristol_grp,met2$Variable)

spear = met2 %>% filter(!is.na(bristol)) %>% 
  group_by(Variable,bristol_grp,Status) %>% 
  group_modify(~cor.test(~value+s__Faecalibacterium_prausnitzii,data=.,method='spearman') %>%
                 .[c('estimate','p.value')] %>% as.data.frame()) %>%
  rename(Rho = estimate) %>% 
  mutate(pval = ifelse(p.value>0.05,'',
                      ifelse(p.value>0.01,'*',
                             ifelse(p.value>0.001,'**',
                                    ifelse(p.value<=0.001,'***',''))))) %>% 
  mutate(pval2 = ifelse(p.value>0.1,'',
                       ifelse(p.value>0.05,'+','')))
spear %>% 
  ggplot(aes(bristol_grp,Variable,col=Rho)) +
  geom_point(size=14) +
  theme_classic(base_size=16) +
  geom_text(aes(label=pval),size = 10, col = 'white',nudge_y = -0.2) +
  geom_text(aes(label=pval2),size = 10, col = 'white',nudge_y = 0) +
  scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
  facet_wrap('Status') +
  xlab('Stool Firmness') + ylab(NULL) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))

ggsave('../Final Outputs/Amino Acid Metabolites/Faecalibacterium_bristolgrp_spearman_apr23.png',height=3,width=6.5)
```

```{r GLM}
met2 = met %>% mutate(bristol_grp = ifelse(is.na(bristol),NA,ifelse(bristol<3,'Firm',ifelse(bristol<5,'Normal','Loose')))) %>% 
  mutate_at(vars('pcresol','phenylacetylglutamine','Log2CRP'),function(x)rescale(x))
met2$bristol_grp = factor(met2$bristol_grp, levels=c('Loose','Normal','Firm'))

met2 = met2 %>% pivot_longer(cols = c('pcresol','phenylacetylglutamine','Log2CRP'),
                             names_to = 'Variable', values_to = 'value') %>% 
  filter(!is.na(value))

met2[met2=='pcresol'] = 'p-Cresol'
met2[met2=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
met2[met2=='Log2CRP'] = 'Log 2 CRP'

table(met2$Status,met2$bristol_grp,met2$Variable)

glm = met2 %>% filter(!is.na(bristol)) %>% 
  group_by(Variable,bristol_grp,Status) %>% 
  group_modify(~summary(glm(s__Faecalibacterium_prausnitzii~value+depth+Sex+ent_boolean,data=.))$coefficients %>%
                 as.data.frame() %>% rownames_to_column('temp')) %>%
  filter(temp=='value') %>% 
  mutate(pval = ifelse(`Pr(>|t|)`>0.05,'',
                      ifelse(`Pr(>|t|)`>0.01,'*',
                             ifelse(`Pr(>|t|)`>0.001,'**',
                                    ifelse(`Pr(>|t|)`<=0.001,'***',''))))) %>% 
  mutate(pval2 = ifelse(`Pr(>|t|)`>0.1,'',
                       ifelse(`Pr(>|t|)`>0.05,'+','')))
glm %>% 
  ggplot(aes(bristol_grp,Variable,col=Estimate)) +
  geom_point(size=14) +
  theme_classic(base_size=16) +
  geom_text(aes(label=pval),size = 10, col = 'white',nudge_y = -0.2) +
  geom_text(aes(label=pval2),size = 10, col = 'white',nudge_y = 0) +
  scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
  facet_wrap('Status') +
  xlab('Stool Firmness') + ylab(NULL) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))

ggsave('../Final Outputs/Amino Acid Metabolites/Faecalibacterium_bristolgrp_glm_apr23.png',height=3,width=6.5)
```

```{r Save Stats}
writexl::write_xlsx(list(Spearman = spear,GLM = glm),
                    '../Final Outputs/Amino Acid Metabolites/faecalibacterium_bristol_statistics_apr23.xlsx')
```