---
title: "5. Network Analysis with SPIEC-EASI"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Initialize}
library(phyloseq)
library(SpiecEasi)
library(tidyverse)
library(Matrix)
library(igraph)
library(ggpubr)
```

# Load data

```{r}
ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds")
ps@sam_data$Status = as.factor(ps@sam_data$Status)
ps@sam_data$bristol = as.numeric(ps@sam_data$bristol)
ps@sam_data$Sex = as.factor(ps@sam_data$Sex)
ps@sam_data$ent_boolean = ps@sam_data$entacapone>0
# Sequencing depth
ps@sam_data$depth = sample_sums(ps)
ps = ps %>% rarefy_even_depth(rngseed = 421)

# Only keep species with >0.1 prevalence
rel = ps@otu_table@.Data %>% as.data.frame %>% mutate_all(function(x)x>0) %>% 
  rowSums %>% as.data.frame(col.names = c('n')) %>% 
  filter(`.` > 0.1*nsamples(ps))
ps = prune_taxa(rownames(rel), ps)

# Subset into disease groups for further analysis
pd = ps %>% subset_samples(Status=='PD')
ctrl = ps %>% subset_samples(Status=='Ctrl')
```

```{r Fine-tune the settings to optimize connectivity, eval=FALSE, include=FALSE}
pargs <- list(rep.num=3, seed=10010)

se.mb.ctrl1 <- spiec.easi(ctrl, method='mb', lambda.min.ratio=1e-1,
                          nlambda=50, pulsar.params=pargs)
se.mb.ctrl2 <- spiec.easi(ctrl, method='mb', lambda.min.ratio=1e-2,
                          nlambda=50, pulsar.params=pargs)
se.mb.ctrl3 <- spiec.easi(ctrl, method='mb', lambda.min.ratio=1e-1,
                          nlambda=20, pulsar.params=pargs)
se.mb.ctrl4 <- spiec.easi(ctrl, method='mb', lambda.min.ratio=1e-2,
                          nlambda=20, pulsar.params=pargs)

library(Matrix)
## Create igraph objects

ig.ctrl1   <- adj2igraph(getRefit(se.mb.ctrl1), vertex.attr=list(name=taxa_names(ps)))
ig.ctrl2   <- adj2igraph(getRefit(se.mb.ctrl2), vertex.attr=list(name=taxa_names(ps)))
ig.ctrl3   <- adj2igraph(getRefit(se.mb.ctrl3), vertex.attr=list(name=taxa_names(ps)))
ig.ctrl4   <- adj2igraph(getRefit(se.mb.ctrl4), vertex.attr=list(name=taxa_names(ps)))

library(igraph)

plot_network(ig.ctrl1, ps, type='taxa', color="Class",label = NA, title = '50-0.1',point_size = 2)
plot_network(ig.ctrl2, ps, type='taxa', color="Class",label = NA, title = '50-0.01',point_size = 2)
plot_network(ig.ctrl3, ps, type='taxa', color="Class",label = NA, title = '20-0.1',point_size = 2)
plot_network(ig.ctrl4, ps, type='taxa', color="Class",label = NA, title = '20-0.01',point_size = 2)
```

# Overall Network Graphs

```{r Networks coloured by Phylum}
# pargs <- list(rep.num=50, seed=10010)

# se.mb.pd <- spiec.easi(pd, method='mb', lambda.min.ratio=1e-1,sel.criterion='bstars',
#                        nlambda=100, pulsar.params=pargs)
# saveRDS(se.mb.pd,'../Final Outputs/SPIEC-EASI/spiec_easi_pd.rds')
# se.mb.ctrl <- spiec.easi(ctrl, method='mb', lambda.min.ratio=1e-1,sel.criterion='bstars',
#                          nlambda=100, pulsar.params=pargs)
# saveRDS(se.mb.ctrl,'../Final Outputs/SPIEC-EASI/spiec_easi_ctrl.rds')

se.mb.pd = readRDS('../Final Outputs/SPIEC-EASI/spiec_easi_pd.rds')
se.mb.ctrl = readRDS('../Final Outputs/SPIEC-EASI/spiec_easi_ctrl.rds')

## Create igraph objects for direct plotting
ig.pd     <- adj2igraph(getRefit(se.mb.pd), vertex.attr=list(name=taxa_names(ps)))
ig.ctrl   <- adj2igraph(getRefit(se.mb.ctrl), vertex.attr=list(name=taxa_names(ps)))

## set size of vertex proportional to clr-mean
vsize.pd    <- rowMeans(clr(t(pd@otu_table@.Data), 1))+6
am.coord.pd <- layout.fruchterman.reingold(ig.pd)
vsize.ctrl    <- rowMeans(clr(t(ctrl@otu_table@.Data), 1))+6
am.coord.ctrl <- layout.fruchterman.reingold(ig.ctrl)

# Plot results
p1 = plot_network(ig.pd, ps, type='taxa', color="Phylum",label = NA, title = 'PD',point_size = vsize.pd-4)
p2 = plot_network(ig.ctrl, ps, type='taxa', color="Phylum",label = NA, title = 'Ctrl',point_size = vsize.ctrl-4)
ggpubr::ggarrange(plotlist=list(p1,p2),common.legend = T, legend = 'right')
ggsave('../Final Outputs/SPIEC-EASI/networks_Phylum.png', height = 5, width = 10)
```

# DEGREE ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Extract edge info from each graph}
conn.pd = c(); conn.ctrl = c() # Length of edges
len.pd = c(); len.ctrl = c() # Number of edges
names.pd = c(); names.ctrl = c() # Species

for(i in 1:length(ig.pd)){
  # i=1
  x = ig.pd[[i]]
  names.pd <- c(names.pd,names(x))
  conn.pd <- c(conn.pd,unname(unlist(x)))
  len.pd <- c(len.pd,length(x[[1]]))
}

for(i in 1:length(ig.ctrl)){
  # i=1
  x = ig.ctrl[[i]]
  names.ctrl <- c(names.ctrl,names(x))
  conn.ctrl <- c(conn.ctrl,unname(unlist(x)))
  len.ctrl <- c(len.ctrl,length(x[[1]]))
}

# Combine into table, add taxonomy info
degree.all = tibble(Status = 'PD', Taxon = names.pd, Connections = len.pd) %>% 
  rbind(tibble(Status = 'Ctrl', Taxon = names.ctrl, Connections = len.ctrl)) %>% 
  left_join(ps@tax_table@.Data %>% as.data.frame() %>% rownames_to_column('Taxon'))
```

## Proportion of species with no connections (Degree = 0)

```{r}
degree_0 = degree.all %>% 
  mutate(Connected = ifelse(Connections==0,'No','Yes')) %>% 
  group_by(Status,Connected) %>% tally()

# Chi square test
degree.chi = degree_0 %>% pivot_wider(names_from = Status,values_from = n) %>% column_to_rownames('Connected')

# 12 unconnected in Ctrl, 21 in PD, out of 133 total species
degree.chi
chi = chisq.test(degree.chi); chi

# Plot:
degree_0 %>% ggplot(aes(Status,n,fill = Connected)) +
  geom_col(position = 'stack') +
  theme_classic(base_size = 20) +
  xlab(NULL) +ylab('# Species') +
  # Add p value
  geom_segment(x = 1,xend = 2,y=140,yend=140, size = 1) +
  geom_text(x = 1.5,y = 150, label=chi$p.value %>% signif(2),size =8) +
  ylim(0,155)
ggsave('../Final Outputs/SPIEC-EASI/degree_0.png', height = 5, width = 5.5)
```

## Compare difference in connections between PD and Ctrl networks in pairwise manner

Pairs are per species.

```{r}
degree.wilcox = wilcox.test(Connections~Status, data = degree.all %>% arrange(Species),paired=T,conf.int = T) %>% unlist %>% as.data.frame %>% rownames_to_column('Parameter') %>% mutate(Test = 'Degree')
names(degree.wilcox)[2] = 'Value'

# No overall differences
degree.all %>% ggplot(aes(Status,Connections,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height = 0.5, width = 0.2) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('# Edges Per Species') +
  geom_line(aes(group = Species)) +
  theme(legend.position = 'none') +
  stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = T,size = 8) +
  ylim(-1,23)
ggsave('../Final Outputs/SPIEC-EASI/degree_all.png', height = 5, width = 4.5)

# No phylum-specific differences
degree.all %>% group_by(Phylum) %>% add_tally %>%
  mutate(Phylum = ifelse(n>4,Phylum,'Other')) %>% ungroup() %>% # Phyla with two or fewer taxa will be grouped
  filter(Phylum != 'Other') %>% # Remove other taxa
  ggplot(aes(Status,Connections,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height=0.1,width =0.1) +
  geom_line(aes(group=Species)) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('# Edges Per Species') +
  theme(legend.position = 'none') +
  ylim(-1,25) +
  ggpubr::stat_compare_means(paired = T, comparisons = list(c('PD','Ctrl')), label = 'p.format',size=6) +
  facet_wrap('Phylum')
ggsave('../Final Outputs/SPIEC-EASI/degree_phylum.png', height = 5, width = 5)
```

# BETWEENNESS CENTRALITY ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Calculate Betweenness Centrality}
bw.pd = betweenness(
  ig.pd,
  v = V(ig.pd),
  cutoff = -1
) %>% as.data.frame()

bw.ctrl = betweenness(
  ig.ctrl,
  v = V(ig.ctrl),
  cutoff = -1
) %>% as.data.frame()

bw = tibble(Status='PD',Taxon = rownames(bw.pd),Betweenness = bw.pd$.) %>% 
  rbind(tibble(Status='Ctrl',Taxon = rownames(bw.ctrl),Betweenness = bw.ctrl$.)) %>% 
  left_join(ps@tax_table@.Data %>% as.data.frame %>% rownames_to_column('Taxon'))
```
```{r}
bw.wilcox = wilcox.test(Betweenness~Status, data = bw %>% arrange(Species),paired=T,conf.int = T) %>% unlist %>% as.data.frame %>% rownames_to_column('Parameter') %>% mutate(Test = 'Betweenness')
names(bw.wilcox)[2] = 'Value'

# Strong overall differences
bw %>% ggplot(aes(Status,Betweenness,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height = 0.5, width = 0.2) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Betweenness Centrality') +
  geom_line(aes(group = Species)) +
  theme(legend.position = 'none') +
  stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = T,size = 8) +
  ylim(-10,2000)
ggsave('../Final Outputs/SPIEC-EASI/betweenness_all.png', height = 5, width = 4.5)

# Phylum-specific differences
bw %>% group_by(Phylum) %>% add_tally %>%
  mutate(Phylum = ifelse(n>4,Phylum,'Other')) %>% ungroup() %>% # Phyla with two or fewer taxa will be grouped
  filter(Phylum != 'Other') %>% # Remove other taxa
  ggplot(aes(Status,Betweenness,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height=0.1,width =0.1) +
  geom_line(aes(group=Species)) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Betweenness Centrality') +
  theme(legend.position = 'none') +
  ylim(-10,2000) +
  ggpubr::stat_compare_means(paired = T, comparisons = list(c('PD','Ctrl')), label = 'p.format',size=6) +
  facet_wrap('Phylum')
ggsave('../Final Outputs/SPIEC-EASI/betweenness_phylum.png', height = 5, width = 5.3)
```

# CLOSENESS CENTRALITY ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r}
close.pd = closeness(
  ig.pd,
  vids = V(ig.pd),
  mode = 'all',
  normalized = TRUE,
  cutoff = -1
) %>% as.data.frame %>% rownames_to_column('V') %>% 
  mutate(Status = 'PD', .before = V)

close.ctrl = closeness(
  ig.ctrl,
  vids = V(ig.ctrl),
  mode = 'all',
  normalized = TRUE,
  cutoff = -1
) %>% as.data.frame %>% rownames_to_column('V') %>% 
  mutate(Status = 'Ctrl', .before = V)

close = close.pd %>% rbind(close.ctrl) %>% rename(Taxon = V) %>% 
  left_join(ps@tax_table@.Data %>% as.data.frame %>% rownames_to_column('Taxon'))
names(close)[3] = 'Closeness'

# Change infilite values to 1.5*max for plotting and stats purposes (these are taxa that aren't connected)
infval = max(close$Closeness,na.rm=T)*1.5
close.noinf = close %>% mutate(Closeness = ifelse(is.nan(Closeness),infval,Closeness))

# Remove infinite closeness (they have no edges and therefore have NaN values), as we want to see how those means change
close.noinf2 = close %>% filter(!is.nan(Closeness))
```

```{r}
close.wilcox = wilcox.test(Closeness~Status, data = close.noinf %>% arrange(Species),paired=T,conf.int = T) %>% unlist %>% as.data.frame %>% rownames_to_column('Parameter') %>% mutate(Test = 'Closeness')
names(close.wilcox)[2] = 'Value'

# Strong overall differences
close.noinf %>% 
  mutate(Closeness = log10(Closeness)) %>% 
  ggplot(aes(Status,Closeness,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height = 0, width = 0.2) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Log 10 Closeness Centrality') +
  geom_line(aes(group = Species)) +
  ylim(-1,0.3) +
  theme(legend.position = 'none') +
  stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = T,size = 8)
ggsave('../Final Outputs/SPIEC-EASI/Closeness_all.png', height = 5, width = 4.5)

# No zero-edge nodes
close.noinf2 %>% 
  mutate(Closeness = log10(Closeness)) %>% 
  ggplot(aes(Status,Closeness,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height = 0, width = 0.2) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Log 10 Closeness Centrality (Non-Inf)') +
  geom_line(aes(group = Species)) +
  ylim(-1,0.3) +
  theme(legend.position = 'none') 
  # stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = T,size = 8)
ggsave('../Final Outputs/SPIEC-EASI/Closeness_all_noinf.png', height = 5, width = 4.5)

# Phylum-specific differences
close.noinf %>% group_by(Phylum) %>% add_tally %>%
  mutate(Phylum = ifelse(n>4,Phylum,'Other')) %>% ungroup() %>% # Phyla with two or fewer taxa will be grouped
  filter(Phylum != 'Other') %>% # Remove other taxa
  mutate(Closeness = log10(Closeness)) %>% 
  ggplot(aes(Status,Closeness,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  geom_jitter(height=0,width =0.1) +
  geom_line(aes(group=Species)) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Closeness Centrality') +
  theme(legend.position = 'none') +
  ylim(-1,0.4) +
  ggpubr::stat_compare_means(paired = T, comparisons = list(c('PD','Ctrl')), label = 'p.format',size=6) +
  facet_wrap('Phylum')
ggsave('../Final Outputs/SPIEC-EASI/Closeness_phylum.png', height = 5, width = 5.3)
```

# TAXON-TAXON DISTANCES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This computes the distance between every connected node pair.

```{r Extract distance information}
dist.pd = distances(
  ig.pd,
  mode = "all",
  algorithm = "automatic"
) %>% as.data.frame %>% rownames_to_column('V') %>% 
  pivot_longer(cols = -V, names_to = 'To', values_to = 'Distance') %>% 
  mutate(Status = 'PD', .before = V)

dist.ctrl = distances(
  ig.ctrl,
  mode = "all",
  algorithm = "automatic"
) %>% as.data.frame %>% rownames_to_column('V') %>% 
  pivot_longer(cols = -V, names_to = 'To', values_to = 'Distance') %>% 
  mutate(Status = 'Ctrl', .before = V)

dist = dist.pd %>% rbind(dist.ctrl) %>% 
  mutate(Connected = ifelse(is.infinite(Distance),'No','Yes'),Comparison = paste0(V,':',To)) %>% 
  filter(Distance !=0) %>% filter(V != To) %>% # Remove self associations
  mutate(connection = NA) # Set up for next bit

# Sort each pair of taxa alphabetically - this will allow me to remove duplicates that were created by including reverse directions.
for(i in 1:nrow(dist)){
  # i=2
  s = sort(c(dist$V[i],dist$To[i]))
  dist$connection[i] = paste0(s[1],s[2])
}

# Remove duplicate reads (i.e. we don't need b-a if we have a-b)
dist_all = dist # Useful for plotting per-phylum distances
dist = dist %>% select(Status,Distance,Connected,connection) %>% unique()

# Change infinite distance (there is no path connecting them and therefore have NaN values) to 1.5*max.
infval = max(dist$Distance[!is.infinite(dist$Distance)],na.rm=T)*1.5
dist.noinf = dist %>% mutate(Distance = ifelse(is.infinite(Distance),infval,Distance))
dist_all.noinf = dist_all %>% mutate(Distance = ifelse(is.infinite(Distance),infval,Distance))
```

## Proportion of existing connections vs total possible connections

```{r}
dist_0 = dist %>% group_by(Status,Connected) %>% tally()

# Chi square test
dist.chi = dist_0 %>% pivot_wider(names_from = Status,values_from = n) %>% column_to_rownames('Connected')
sum(dist.chi$Ctrl) # of possible connections
dist.chi
chi = chisq.test(dist.chi); chi

# Plot:
dist_0 %>% ggplot(aes(Status,n,fill = Connected)) +
  geom_col(position = 'stack') +
  theme_classic(base_size = 20) +
  xlab(NULL) +ylab('# Taxon-Taxon Connections') +
  # Add p value
  geom_segment(x = 0.8,xend = 2.2,y=9000,yend=9000, size = 1) +
  geom_text(x = 1.5,y = 9500, label='p < 2.22e-16',size =8) +
  ylim(0,9500)
ggsave('../Final Outputs/SPIEC-EASI/dist_0.png', height = 5, width = 5.5)
```

## Compare difference in connections between PD and Ctrl networks in pairwise manner

Pairs are per species.

```{r}
dist.wilcox = wilcox.test(Distance~Status, data = dist.noinf %>% arrange(connection),paired=T,conf.int = T) %>% unlist %>% as.data.frame %>% rownames_to_column('Parameter') %>% mutate(Test = 'dist')
names(dist.wilcox)[2] = 'Value'

dist.noinf %>% ggplot(aes(Status,Distance,fill = Status)) +
  geom_violin(adjust = 2, draw_quantiles = 0.5,size = 1) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Taxon-Taxon Distance') +
  theme(legend.position = 'none') +
  stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = T,size = 8) +
  ylim(0,20)
ggsave('../Final Outputs/SPIEC-EASI/dist_all.png', height = 5, width = 4.5)

dist.noinf %>% 
  filter(Distance<18) %>% 
  ggplot(aes(Status,Distance,fill = Status)) +
  geom_violin(adjust = 3
              , draw_quantiles = 0.5,size = 1) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Taxon-Taxon Distance (Non-Inf)') +
  theme(legend.position = 'none') +
  stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = F,size = 8) +
  ylim(0,14)
ggsave('../Final Outputs/SPIEC-EASI/dist_all_noninf.png', height = 5, width = 4.5)
```

```{r}
temp = dist_all.noinf

temp %>% 
  filter(Distance<18) %>% 
  ggplot(aes(Status,Distance,fill = Status)) +
  geom_violin(adjust = 3
              , draw_quantiles = 0.5,size = 1) +
  theme_classic(base_size = 20) +
  xlab(NULL) + ylab('Taxon-Taxon Distance (Non-Inf)') +
  theme(legend.position = 'none') +
  stat_compare_means(comparisons = list(c('PD','Ctrl')),label = 'p.format', paired = F,size = 8) +
  ylim(0,14) + facet_wrap('Phylum')
ggsave('../Final Outputs/SPIEC-EASI/dist_all_noninf.png', height = 5, width = 4.5)
```

```{r}
phylum_list = c("p__Actinobacteria","p__Bacteroidetes","p__Firmicutes","p__Proteobacteria")
dist_all_0 = dist_all %>% rename(Taxon = V) %>% 
  left_join(ps@tax_table@.Data %>% as.data.frame() %>% rownames_to_column('Taxon')) %>%
  filter(Phylum %in% phylum_list) %>% 
  group_by(Phylum,Status,Connected) %>% tally() %>% ungroup

# Chi square test
pvals = c()
for(i in phylum_list){
  print(i)
  dist.chi = dist_all_0 %>% filter(Phylum==i) %>% select(-Phylum) %>% 
    pivot_wider(names_from = Status,values_from = n) %>% column_to_rownames('Connected')
  sum(dist.chi$Ctrl) # of possible connections
  dist.chi
  chi = chisq.test(dist.chi); pvals = c(pvals,chi$p.value); print(chi$p.value)
}

# Plot:
dist_all_0 %>% 
  left_join(tibble(Phylum = phylum_list, pvals = scales::scientific(pvals,3))) %>% 
  group_by(Status,Phylum) %>% mutate(n = n/sum(n)) %>% 
  ggplot(aes(Status,n,fill = Connected)) +
  geom_col(position = 'stack') +
  theme_classic(base_size = 20) +
  xlab(NULL) +ylab('Prop. Possible Connections') +
  # Add p value
  geom_segment(x = 0.8,xend = 2.2,y=1.1,yend=1.1, size = 1) +
  geom_text(x = 1.5,y = 1.3,aes(label=pvals),size =7) +
  ylim(0,1.4) +
  facet_wrap('Phylum')
ggsave('../Final Outputs/SPIEC-EASI/dist_0_phylum.png', height = 5, width = 7)
```