---
title: "5. Species with vs Metabolites and Disease Progression"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load data}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(microbiome)
library(ggcorrplot)
library(lme4)
library(lmerTest)
library(scales)

source('Supp Scripts/5.1. Load Datasets.R')
source('Supp Scripts/5.3. LMER Functions.R')
source('Supp Scripts/5.4. Cross-Sectional Functions.R')
```

```{r Metabolite (amino acid) data}
# Other data
aa = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/Reference Files/Metadata/aa_metabolites.csv') %>% 
  rename(ID = Record.ID) %>% mutate(ID = as.character(ID))
id = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/Reference Files/ID_key.csv')

met = aa %>% rename(record_id = ID) %>% 
  left_join(id)
met$record_id[!is.na(met$correct_id)] = met$correct_id[!is.na(met$correct_id)]
met = met %>% rename(Sample = record_id) %>% select(-correct_id)
met$Sample = paste('X_',met$Sample,sep='')
rm(aa,id)

# Join taxonomy data
met = met %>% left_join(temp) %>% 
  # Remove anyone without taxonomy data
  filter(!is.na(s__Ruthenibacterium_lactatiformans))
```

```{r Longitudinal Data}
# Load dataset, choose relevant metadata
long0 = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/Reference Files/Metadata/shayanandkiradata_nov22.csv') %>% 
  select(Record.ID,redcap_event_name,Disease.duration,bdi_total,contains('mds'),td.pigd,fss_total,months_last_visit) %>% 
  rename(Sample = Record.ID) %>% 
  # Add an X to sample names to match the other datasets
  mutate(Sample = paste0('X_',Sample)) %>% 
  arrange(months_last_visit) %>% 
  group_by(Sample) %>% 
  # Set the baseline as 0 months
  mutate(months_last_visit = ifelse(redcap_event_name=='visit_1_arm_1',0,months_last_visit)) %>%
  # Select only those with longitudinal data available
  add_count() %>% filter(n>1) %>% select(-n) %>%
  ungroup

# Select microbiome-relevant metadata, join to longitudinal dataset
micro = psm %>% select(Sample,Status,Sex,bristol,ent_boolean,depth,OTU,Abundance,Kingdom:Species) %>% unique
long = long0 %>% left_join(micro)
```

```{r Average study duration}
test = long %>% filter(Status=='PD',!is.na(months_last_visit)) %>% 
  select(-c(OTU,Abundance,Kingdom:Species)) %>% unique()
test2 = test %>% arrange(months_last_visit) %>% 
  group_by(Sample) %>% 
  filter(row_number()==n())

mean(test2$months_last_visit); sd(test2$months_last_visit)
```

# *We now have two datasets: met (metabolomics) and long (longitudinal analysis)*

Both have clr-transformed species of interest attached.

'sig' can be used as a vector of all significant species.

# PART 1: CROSS-SECTIONAL SPECIES VS METABOLITES~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Ctrl and PD tested separately and together.

met used for amino acids, crpbut used for crp and butyrate.

```{r Spearman associations of metabolites with species}
spearman = correlate_met(met,'pcresol',meth='spearman',grp='All') %>% 
  # metabolite data
  rbind(correlate_met(met,'phenylacetylglutamine',meth='spearman',grp='All')) %>% 
  rbind(correlate_met(met %>% filter(Status=='PD'),'pcresol',meth='spearman',grp='PD')) %>% 
  rbind(correlate_met(met %>% filter(Status=='PD'),'phenylacetylglutamine',meth='spearman',grp='PD')) %>% 
  rbind(correlate_met(met %>% filter(Status=='Ctrl'),'pcresol',meth='spearman',grp='Ctrl')) %>% 
  rbind(correlate_met(met %>% filter(Status=='Ctrl'),'phenylacetylglutamine',meth='spearman',grp='Ctrl')) %>% 
  # crpbut contains all samples, not just those with metabolite data
  rbind(correlate_met(crpbut,'Log2CRP',meth='spearman',grp='All')) %>% 
  rbind(correlate_met(crpbut,'butyrate',meth='spearman',grp='All')) %>% 
  rbind(correlate_met(crpbut %>% filter(Status=='PD'),'Log2CRP',meth='spearman',grp='PD')) %>% 
  rbind(correlate_met(crpbut %>% filter(Status=='PD'),'butyrate',meth='spearman',grp='PD')) %>% 
  rbind(correlate_met(crpbut %>% filter(Status=='Ctrl'),'Log2CRP',meth='spearman',grp='Ctrl')) %>% 
  rbind(correlate_met(crpbut %>% filter(Status=='Ctrl'),'butyrate',meth='spearman',grp='Ctrl'))
```

```{r Plot Spearman Metabolites}
# Edit levels for plotting
spearman[spearman=='pcresol'] = 'p-Cresol'
spearman[spearman=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
spearman[spearman=='butyrate'] = 'Butyrate Synth.\nCapacity'
spearman[spearman=='Log2CRP'] = 'Log 2 CRP'
spearman$taxon = str_remove_all(spearman$taxon,'s__') %>% str_replace_all('_',' ')

prep_plot = function(vars){
  p = spearman %>% 
    filter(variable %in% c(vars)) %>% 
    rename("Rho" = rho) %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    mutate(variable = ifelse(variable=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',variable)) %>% 
    ggplot(aes(group,taxon,col = Rho)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 10, col = 'white',nudge_y = -0.25) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('variable')
  return(p)
}

# Arrange significance by p-cresol, all estimates ~~~~~~~~~~~~~
temp = spearman %>% filter(group=='All',variable=='p-Cresol') %>% 
  arrange(rho)
spearman$taxon = factor(spearman$taxon, levels = temp$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))

# Arrange significance by Butyrate Synth.\nCapacity, all estimates~~~~~~~~~~~~~
temp = spearman %>% filter(group=='All',variable=='Butyrate Synth.\nCapacity') %>% 
  arrange(rho)
spearman$taxon = factor(spearman$taxon, levels = temp$taxon)

# Plot butyrate
p2 = prep_plot('Butyrate Synth.\nCapacity')

# Arrange significance by Log 2 CRP, all estimates ~~~~~~~~~~~~~~~~
temp = spearman %>% filter(group=='All',variable=='Log 2 CRP') %>% 
  arrange(rho)
spearman$taxon = factor(spearman$taxon, levels = temp$taxon)
# Plot CRP
p3 = prep_plot('Log 2 CRP')

ggarrange(plotlist = list(p1,p2,p3),ncol = 3, nrow=1,widths=c(4,3,3))
ggsave('../Final Outputs/Amino Acid Metabolites/aa_but_crp_vs_sig_taxa_heatmap_v2.png',height=6,width=19)
```

```{r LM associations of metabolites with species}

# Rescaling species and metabolites to 0-1 before running so that all estimates are comparable!

glm = glm_met(met,'pcresol',grp='All') %>% 
  rbind(glm_met(met,'phenylacetylglutamine',grp='All')) %>% 
  rbind(glm_met(met %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(glm_met(met %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(glm_met(met %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(glm_met(met %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl')) %>% 
  rbind(glm_met(crpbut,'butyrate',grp='All')) %>% 
  rbind(glm_met(crpbut,'Log2CRP',grp='All')) %>% 
  rbind(glm_met(crpbut %>% filter(Status=='PD'),'butyrate',grp='PD')) %>% 
  rbind(glm_met(crpbut %>% filter(Status=='PD'),'Log2CRP',grp='PD')) %>% 
  rbind(glm_met(crpbut %>% filter(Status=='Ctrl'),'butyrate',grp='Ctrl')) %>% 
  rbind(glm_met(crpbut %>% filter(Status=='Ctrl'),'Log2CRP',grp='Ctrl'))

glm_species = glm %>% filter(str_detect(Explanatory,'s__'))
```

```{r Plot glm_species Metabolites}
# Edit levels for plotting
glm_species[glm_species=='pcresol'] = 'p-Cresol'
glm_species[glm_species=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
glm_species[glm_species=='butyrate'] = 'Butyrate Synth.\nCapacity'
glm_species[glm_species=='Log2CRP'] = 'Log 2 CRP'
glm_species$Explanatory = str_remove_all(glm_species$Explanatory,'s__') %>% str_replace_all('_',' ')

prep_plot = function(vars){
  p = glm_species %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(Response %in% vars) %>% 
    mutate(Response = ifelse(Response=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',Response)) %>% 
    ggplot(aes(group,Explanatory,col = Estimate)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 10, col = 'white',nudge_y = -0.25) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('Response')
  return(p)
}

#Arrange significance by p-cresol, all estimates ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
temp = glm_species %>% filter(group=='All',Response=='p-Cresol') %>% 
  arrange(Estimate)
glm_species$Explanatory = factor(glm_species$Explanatory, levels = temp$Explanatory)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))

#Arrange significance by butyrate, all estimates ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
temp = glm_species %>% filter(group=='All',Response=='Butyrate Synth.\nCapacity') %>% 
  arrange(Estimate)
glm_species$Explanatory = factor(glm_species$Explanatory, levels = temp$Explanatory)

# Plot
p2 = prep_plot('Butyrate Synth.\nCapacity')

#Arrange significance by CRP, all estimates ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
temp = glm_species %>% filter(group=='All',Response=='Log 2 CRP') %>% 
  arrange(Estimate)
glm_species$Explanatory = factor(glm_species$Explanatory, levels = temp$Explanatory)

# Plot
p3 = prep_plot('Log 2 CRP')

ggarrange(plotlist = list(p1,p2,p3),ncol = 3, nrow=1,widths=c(4,3,3))
ggsave('../Final Outputs/Amino Acid Metabolites/aa_but_crp_vs_sig_taxa_heatmap_glm_v2.png',height=6,width=19)
```

```{r Save Stats}
writexl::write_xlsx(list(Spearman = spearman,GLM = glm,'GLM Species Only' = glm_species),
                    '../Final Outputs/Amino Acid Metabolites/amino_acid_but_crp_metabolite_statistics.xlsx')
```


# PART 2 LONGITUDINAL SEVERITY ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r set up}
# Center static predictors, normalize as necessary. This reduces spurious correlations as directed in documentation.
# Center each OTU separately
pd = long %>% filter(Status == 'PD') %>% group_by(OTU) %>% 
  mutate(Abundance = rescale(Abundance)) %>% 
  # Rescale to range of 0-1
  mutate(Abundance = Abundance-mean(Abundance,na.rm=T)) %>% ungroup()
pd$months_last_visit = log10(pd$months_last_visit + 1) # Log10 transform before centering
pd$months_last_visit = pd$months_last_visit - mean(pd$months_last_visit,na.rm=T)
pd$bristol = pd$bristol - mean(pd$bristol,na.rm=T)
pd$depth = (pd$depth - mean(pd$depth,na.rm=T))/1e6 # Divide to reduce magnitude as directed in documentation
pd$Sex = factor(pd$Sex)
pd$ent_boolean = factor(pd$ent_boolean)
```

## ALL SAMPLES

```{r Run LMER}
vars = c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','fss_total','bdi_total')

# Univariate
allres = data.frame()
for(v in vars) { for(i in df$sig) run_lmer(pd,i,v) }

# Multivariate - Interaction (no DD)
allresmulti = data.frame()
for(v in vars) { for(i in df$sig) run_lmer_multi(pd,i,v) }
allresmulti.int = allresmulti

# Multivariate - Abundance (incl DD)
allresmulti = data.frame()
for(v in vars) { for(i in df$sig) run_lmer_multi2(pd,i,v) }
allresmulti.ab = allresmulti

# Process results
allres.p = allres %>% filter(Variable == 'Abundance:months_last_visit') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()

allres.ab.p = allres %>% filter(Variable == 'Abundance') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()

allresmulti.int.p = allresmulti.int %>% filter(Variable == 'Abundance:months_last_visit') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()

allresmulti.ab.p = allresmulti.ab %>% filter(Variable == 'Abundance') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()
```

```{r Plot LMER}
int.uni = plot_lmer(allres.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
int.multi = plot_lmer(allresmulti.int.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
ab.uni = plot_lmer(allres.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))
ab.multi = plot_lmer(allresmulti.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))

ggsave(plot = int.uni,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_int.uni.png',height=6,width=7.5)
ggsave(plot = int.multi,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_int.multi.png',height=6,width=7.5)
ggsave(plot = ab.uni,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_ab.uni.png',height=6,width=8.5)
ggsave(plot = ab.multi,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_ab.multi.png',height=6,width=8.5)
```

```{r Save Stats}
int.uni = edit_lmer(allres.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
int.multi = edit_lmer(allresmulti.int.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
ab.uni = edit_lmer(allres.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))
ab.multi = edit_lmer(allresmulti.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))

writexl::write_xlsx(list('Interaction Univar' = int.uni, 'Interaction Multivar' = int.multi,
                         'Abundance Univar' = ab.uni, 'Abundance Multivar' = ab.multi),
                    '../Final Outputs/Clinical Data/lmer_statistics_species.xlsx')
```

## TD VS PIGD

```{r set up}
# Categorize - must have same class throughout
cat = pd %>% select(Sample, td.pigd) %>%
  filter(!is.na(td.pigd)) %>% 
  mutate(td.pigd = ifelse(td.pigd>=1.5,"TD",ifelse(td.pigd>1,'Mixed','PIGD'))) %>% unique %>% 
  group_by(Sample) %>% add_count() %>% ungroup() %>% filter(n==1)

pd.td = pd %>% filter(Sample %in% cat$Sample[cat$td.pigd=='TD'])
pd.pigd = pd %>% filter(Sample %in% cat$Sample[cat$td.pigd=='PIGD'])

table(cat$td.pigd)
```

```{r Run LMER}
vars = c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','fss_total','bdi_total')

# TD ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Univariate
allres = data.frame()
for(v in vars) { for(i in sig) run_lmer(pd.td,i,v) }
# Multivariate - Interaction (no DD)
allresmulti = data.frame()
for(v in vars) { for(i in sig) run_lmer_multi(pd.td,i,v) }
allresmulti.int = allresmulti
# Multivariate - Abundance (incl DD)
allresmulti = data.frame()
for(v in vars) { for(i in sig) run_lmer_multi2(pd.td,i,v) }
allresmulti.ab = allresmulti

allres.td = allres; allresmulti.int.td = allresmulti.int; allresmulti.ab.td = allresmulti.ab

# PIGD ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Univariate
allres = data.frame()
for(v in vars) { for(i in sig) run_lmer(pd.pigd,i,v) }
# Multivariate - Interaction (no DD)
allresmulti = data.frame()
for(v in vars) { for(i in sig) run_lmer_multi(pd.pigd,i,v) }
allresmulti.int = allresmulti
# Multivariate - Abundance (incl DD)
allresmulti = data.frame()
for(v in vars) { for(i in sig) run_lmer_multi2(pd.pigd,i,v) }
allresmulti.ab = allresmulti

allres.pigd = allres; allresmulti.int.pigd = allresmulti.int; allresmulti.ab.pigd = allresmulti.ab

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# COMBINE
allres = rbind(allres.td %>% mutate(td.pigd = 'TD'), allres.pigd %>% mutate(td.pigd = 'PIGD'))
allresmulti.int = rbind(allresmulti.int.td %>% mutate(td.pigd = 'TD'), allresmulti.int.pigd %>% mutate(td.pigd = 'PIGD'))
allresmulti.ab = rbind(allresmulti.ab.td %>% mutate(td.pigd = 'TD'), allresmulti.ab.pigd %>% mutate(td.pigd = 'PIGD'))

# Process results
allres.p = allres %>% filter(Variable == 'Abundance:months_last_visit') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test,td.pigd) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()

allres.ab.p = allres %>% filter(Variable == 'Abundance') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test,td.pigd) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()

allresmulti.int.p = allresmulti.int %>% filter(Variable == 'Abundance:months_last_visit') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test,td.pigd) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()

allresmulti.ab.p = allresmulti.ab %>% filter(Variable == 'Abundance') %>% arrange(`Pr(>|t|)`) %>% 
  group_by(Test,td.pigd) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()
```

```{r Plot LMER}
int.uni = plot_lmer2(allres.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
int.multi = plot_lmer2(allresmulti.int.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
ab.uni = plot_lmer2(allres.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))
ab.multi = plot_lmer2(allresmulti.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))

ggsave(plot = int.uni,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_int.uni - tdpigd.png',height=10.5,width=8)
ggsave(plot = int.multi,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_int.multi - tdpigd.png',height=10.5,width=8)
ggsave(plot = ab.uni,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_ab.uni - tdpigd.png',height=10.5,width=10)
ggsave(plot = ab.multi,'../Final Outputs/Clinical Data/long_vs_sig_taxa_heatmap_ab.multi - tdpigd.png',height=10.5,width=10)
```

```{r Save Stats}
int.uni = edit_lmer(allres.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
int.multi = edit_lmer(allresmulti.int.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total'))
ab.uni = edit_lmer(allres.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))
ab.multi = edit_lmer(allresmulti.ab.p,c('mds.total','mds1.total','mds2.total','mds3.total','mds4.total','bdi_total','fss_total'))

writexl::write_xlsx(list('Interaction Univar' = int.uni, 'Interaction Multivar' = int.multi,
                         'Abundance Univar' = ab.uni, 'Abundance Multivar' = ab.multi),
                    '../Final Outputs/Clinical Data/lmer_statistics_species_tdpigd.xlsx')
```


# PART 3 AMINO ACIDS VS LONGITUDINAL DATA ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r}
ml = long %>% select(-c(OTU,Kingdom:Species, Abundance)) %>% unique() %>% 
  left_join(met %>% select(Sample:pcresol)) %>% filter(Status=='PD', !is.na(pcresol)) %>% 
  pivot_longer(cols = c(pcresol,phenylacetylglutamine), names_to = 'OTU', values_to = 'Abundance')

# Center static predictors, normalize as necessary. This reduces spurious correlations as directed in documentation.
# Center each OTU separately
pd = ml %>% group_by(OTU) %>% 
  mutate(Abundance = (Abundance-mean(Abundance,na.rm=T))/1e5) %>% ungroup() # Divide to reduce magnitude as directed in documentation
pd$months_last_visit = log10(pd$months_last_visit + 1) # Log10 transform before centering
pd$months_last_visit = pd$months_last_visit - mean(pd$months_last_visit,na.rm=T)
pd$bristol = pd$bristol - mean(pd$bristol,na.rm=T)
pd$depth = (pd$depth - mean(pd$depth,na.rm=T))/1e6 # Divide to reduce magnitude as directed in documentation
pd$Sex = factor(pd$Sex)
pd$ent_boolean = factor(pd$ent_boolean)

hist(pd$Abundance)

allres = data.frame()
run_lmer = function(df,species,variable,log=F,pc = 1, meanscale=T){
  names(df)[which(names(df)==variable)] = 'test'
  if(log){df$test = log10(df$test+pc)}
  if(meanscale){df$test = df$test-mean(df$test,na.rm=T)}
  a = lmer(test ~ Abundance*months_last_visit + (months_last_visit|Sample), 
           data = df %>% filter(OTU==species), na.action = 'na.omit',REML=F, 
           control = lmerControl(optimizer="bobyqa", optCtrl=list(maxfun=100000))) %>% 
    summary %>% .$coefficients %>% as.data.frame() %>% rownames_to_column('Variable') %>% mutate(Taxon = species,Test=variable)
  a$Variable = sapply(a$Variable,function(x)str_replace_all(x,'test',variable))
  allres <<- allres %>% rbind(a)
}

for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd,i,'mds.total')}
for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd,i,'mds1.total')}
for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd,i,'mds2.total')}
for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd,i,'mds3.total')}
for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd,i,'mds4.total')}
# for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd %>% filter(!is.na(fss_total)),i,'fss_total')}
# for(i in c('pcresol','phenylacetylglutamine')){ run_lmer(pd,i,'bdi_total')}

allres2 = allres %>% arrange(`Pr(>|t|)`,Variable) %>% 
  group_by(Test) %>% mutate(qval = p.adjust(`Pr(>|t|)`,method = 'fdr')) %>% ungroup()
```

```{r save stats}
writexl::write_xlsx(list(allres2),
                    '../Final Outputs/Clinical Data/disease_progression_aminoacid_statistics.xlsx')
```