---
title: "6. Diversity of Functional Data"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(phyloseq)
library(ggpubr)
library(data.table)
library(vegan)
library(tidyverse)
gc()

source('../id_functions.R')

prep_tsv = function(df){
  names(df)[1] = 'temp'
  return(df %>% filter(!(temp %in% c('UNMAPPED','UNINTEGRATED','UNGROUPED'))) %>% column_to_rownames('temp') %>%
           select(-contains('Atcc'),-contains('Zymo'),-contains('blnk')) %>%
           `colnames<-`(reformat_ids(colnames(.)) %>% change_ids_to_longitudinal()))
}
# Load RPK Data~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
df1 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_eggnog_renamed_unstratified.tsv') %>% prep_tsv()
df2 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_ko_renamed_unstratified.tsv') %>% prep_tsv()
df3 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_rxn_renamed_unstratified.tsv') %>% prep_tsv()
df4 = fread('../../../Metagenomics/RPK counts/merged_pathabundance_unstratified.tsv') %>% prep_tsv()

# Add to 1 list for high-throughput coding
df = list(eggnog = df1, ko = df2, rxn = df3, pwy = df4)

# Load total metagenomic reads~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cnt = read_delim('../../../Metagenomics/ORIGINAL DATA/mqc_fastqc_sequence_counts_plot_1.txt') %>%
  filter(!str_detect(Sample,'Atcc')&!str_detect(Sample,'blnk')&!str_detect(Sample,'Zymo')) %>% 
  filter(str_detect(Sample,'bmtagged_1')) %>% 
  mutate(Sample = reformat_ids(Sample) %>% change_ids_to_longitudinal()) %>%
  mutate(depth=`Unique Reads`+`Duplicate Reads`) %>%
  select(Sample,depth)

# Load metadata and convert to phyoseq object~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds")
s = psmelt(ps) %>% select(-OTU,-Abundance,-any_of(c('Kingdom','Phylum','Class','Order',
                                                'Family','Genus','Species'))) %>% unique() %>% 
  mutate(Status = as.factor(Status),
         bristol = as.numeric(bristol),
         Sex = as.factor(Sex),
         ent_boolean = entacapone>0) %>% 
  left_join(cnt) %>% `rownames<-`(.$Sample)

# Taxonomy tables
t = lapply(df,function(x) matrix(data = rownames(x), ncol = 1, dimnames = list(rownames(x),'Species')))

# Filter OTU tables to remove the two failed samples (no bacterial OTUs)
df = df %>% lapply(function(x) x %>% select(all_of(rownames(s))) %>% as.matrix)

# Create phyloseq objects
ps = list()
for(i in 1:4){
  temp = phyloseq(sample_data(s),otu_table(df[[i]],taxa_are_rows = T),tax_table(t[[i]]))
  ps = ps %>% append(temp)
}
names(ps) = names(df)

# Filter enzyme data to only include COG dataset
ps$eggnog = ps$eggnog %>% subset_taxa(str_sub(Species,end=3)=='COG') # COG only

# Round counts and rarefy
set.seed(421)
ps = lapply(ps,function(x) transform_sample_counts(x,function(y)round(y,0)) %>% rarefy_even_depth)

# All done the formatting - everything is now nicely arranged inside a list of phyloseq objects!
rm(cnt,df1,df2,df3,df4,t,temp,i,prep_tsv)
```

# DIVERSITY ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Set up}
otulist = lapply(ps,function(x) x@otu_table %>% as.matrix %>% as.data.frame)
```

## ALPHA DIVERSITY ~~~~~~~~~~~~~~~~~~~~~

Using Shannon for alpha diversity.

```{r Alpha Diversity}
# This function will be run on each of the functional datasets.
run_alpha = function(x){
  # Use Vegan package to calculate diversity
  df_alpha <- diversity(t(x), index='shannon', MARGIN = 1, base = exp(1)) %>%
    as.data.frame()
  df_alpha$Sample = rownames(df_alpha)
  df_alpha = df_alpha %>% left_join(s %>% select(Status,depth,Sex,bristol,ent_boolean) %>% rownames_to_column('Sample'))
  names(df_alpha)[1] = 'shannon'
    
  # Plot results
  p = ggplot(df_alpha, aes(x=Status, y=shannon)) + 
        geom_boxplot(outlier.shape = NA) +
        geom_jitter(aes(color=Status),  height = 0, width = 0.2) + 
        theme_classic(base_size = 20) +
        stat_compare_means(comparisons = list(c('PD','Ctrl')),size = 5) +  xlab(NULL) +
        theme(legend.position = 'none') +
        scale_y_continuous(expand = expansion(mult = 0.1))
  
  g = summary(glm(shannon~Status+Sex+bristol+ent_boolean+depth,data = df_alpha))$coefficients
  w = wilcox.test(shannon~Status, data = df_alpha) %>% broom::tidy()

  plotlist <<- plotlist %>% append(list(p))
  alphalist <<- alphalist %>% append(list(df_alpha))
  glmlist <<- glmlist %>% append(list(g))
  wilcoxlist <<- wilcoxlist %>% append(list(w))
}

# Prep some empty lists
plotlist = list()
alphalist = list()
glmlist = list()
wilcoxlist = list()

# Run alpha diversity analysis
lapply(otulist,run_alpha)

# Edit outputs
names(plotlist) = c('COG Enzymes','KEGG','EC','MetaCyc')
names(alphalist) = c('COG Enzymes','KEGG','EC','MetaCyc')
names(glmlist) = c('COG Enzymes','KEGG','EC','MetaCyc')

# Add descriptive y axes
for(i in names(plotlist)){
  plotlist[[i]]$labels$y = i
}

# Combine and save plots
ggarrange(plotlist = plotlist, ncol = 2, nrow = 2)
ggsave(filename='../Final Outputs/Diversity/alpha_diversity_function.png',height = 7, width = 6)

# Format multivariate results
glmlist = glmlist %>% lapply(function(x) x %>% as.data.frame() %>% rownames_to_column('Variable'))
for(i in 1:4){glmlist[[i]] = glmlist[[i]] %>% mutate(Type = names(glmlist)[i])}

# Output of all GLM data
glmall = bind_rows(glmlist) %>% select(Type,everything()) %>% 
  group_by(Variable) %>% mutate(qval = p.adjust(`Pr(>|t|)`)) %>% ungroup() %>% arrange(qval)

# Output of all Wilcoxon data
wilall = bind_rows(wilcoxlist) %>% mutate() %>% 
  group_by(Variable) %>% mutate(qval = p.adjust(`Pr(>|t|)`)) %>% ungroup() %>% arrange(qval)
```

## BETA DIVERSITY ~~~~~~~~~~~~~~~~~~~~~

Using Bray-Curtis for beta diversity.

```{r Beta Univariate}
run_beta = function(x){
  # x = otulist$pwy
  # Calculate the distance matrix:
  beta_dist <- vegdist(t(x), index = 'bray')
  # Scaling results with NMDS
  mds <- metaMDS(beta_dist)
  # Extract PCoA axes and format
  mds_data <- as.data.frame(mds$points)
  mds_data$`Sample` <- rownames(mds_data)
  mds_data <- mds_data %>% left_join(s %>% select(Status,depth,Sex,bristol,ent_boolean) %>% rownames_to_column('Sample')) %>% 
    mutate(beta_type = 'Bray-Curtis')
  
  # Plot results
  p = ggplot(mds_data, aes(x = MDS1, y = MDS2, color = Status)) +
      geom_point(size = 2) + 
      stat_ellipse(size = 1) +
      theme_classic(base_size = 20) + 
      xlab('PCoA 1') + ylab('PCoA 2') +
      ggtitle('Bray-Curtis Beta Diversity')
  
  # Univariate Statistics
  g1 = adonis2(beta_dist ~ Status, data = mds_data)
  md2 = mds_data %>% filter(!is.na(bristol),!is.na(ent_boolean),!is.na(Sex),!is.na(depth))
  bd2 = t(x);bd2 = bd2[rownames(bd2) %in% md2$Sample,]
  g2 = adonis2(bd2 ~ Status+Sex+bristol+depth+ent_boolean, 
               data = md2, method = 'bray', by='margin')
  
  plotbeta <<- plotbeta %>% append(list(p))
  betalist <<- betalist %>% append(list(mds_data))
  glmbeta <<- glmbeta %>% append(list(g1))
  glmbeta.m <<- glmbeta.m %>% append(list(g2))
}

plotbeta = list()
betalist = list()
glmbeta = list()
glmbeta.m = list()
lapply(otulist,run_beta)
names(plotbeta) = c('COG Enzymes','KEGG','EC','MetaCyc')
names(betalist) = c('COG Enzymes','KEGG','EC','MetaCyc')
names(glmbeta) = c('COG Enzymes','KEGG','EC','MetaCyc')
names(glmbeta.m) = c('COG Enzymes','KEGG','EC','MetaCyc')

# Add descriptive y axes
for(i in names(plotbeta)){
  plotbeta[[i]]$labels$title = i
}

# Combine and save plots
ggarrange(plotlist = plotbeta, ncol = 2, nrow = 2, legend = 'right',common.legend = T)
ggsave(filename='../Final Outputs/Diversity/beta_diversity_function.png',height = 7, width = 8)

# Beta Statistics

glmbeta = glmbeta %>% lapply(function(x) x %>% as.data.frame() %>% rownames_to_column('Variable'))
glmbeta.m = glmbeta.m %>% lapply(function(x) x %>% as.data.frame() %>% rownames_to_column('Variable'))
for(i in 1:length(glmbeta)){
  glmbeta[[i]] = glmbeta[[i]] %>% mutate(Type = names(glmbeta)[i])
  glmbeta.m[[i]] = glmbeta.m[[i]] %>% mutate(Type = names(glmbeta.m)[i])
}

# Output of all GLM data
glmbetaall = bind_rows(glmbeta) %>% select(Type,everything()) %>% filter(!(Variable %in% c('Residual','Total'))) %>% 
  mutate(qval = p.adjust(`Pr(>F)`,method='fdr') %>% round(3)) %>% arrange(`Pr(>F)`)
glmbetaall.m = bind_rows(glmbeta.m) %>% select(Type,everything()) %>% filter(!(Variable %in% c('Residual','Total'))) %>% 
  group_by(Variable) %>% 
  mutate(qval = p.adjust(`Pr(>F)`,method='fdr') %>% round(3)) %>% ungroup() %>% arrange(`Pr(>F)`)
```

# SAVE STATS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r}
writexl::write_xlsx(list('Shannon Diversity' = glmall,
                         'Bray-Curtis Univariate' = glmbetaall, 'Bray-Curtis Multivariate' = glmbetaall.m),
                    '../Supplementary Tables/6. Diversity of Functional Data.xlsx')
```

