---
title: "2. Maaslin2 of Functional Data"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(phyloseq)
library(Maaslin2)
library(ggpubr)
library(data.table)
library(vegan)
library(tidyverse)

source("C:/Users/armetcal/OneDrive - UBC/Grad School/CODES/R Functions/da_supplementary_functions_v4.R")
source('../id_functions.R')

prep_tsv = function(df){
  names(df)[1] = 'temp'
  return(df %>% filter(!(temp %in% c('UNMAPPED','UNINTEGRATED','UNGROUPED'))) %>% column_to_rownames('temp') %>%
           select(-contains('Atcc'),-contains('Zymo'),-contains('blnk')) %>%
           `colnames<-`(reformat_ids(colnames(.)) %>% change_ids_to_longitudinal()))
}
# Load RPK Data~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
df1 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_eggnog_renamed_unstratified.tsv') %>% prep_tsv()
df2 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_ko_renamed_unstratified.tsv') %>% prep_tsv()
df3 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_rxn_renamed_unstratified.tsv') %>% prep_tsv()
df4 = fread('../../../Metagenomics/RPK counts/merged_pathabundance_unstratified.tsv') %>% prep_tsv()

# Add to 1 list for high-throughput coding
df = list(eggnog = df1, ko = df2, rxn = df3, pwy = df4)

# Load total metagenomic reads~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cnt = read_delim('../../../Metagenomics/ORIGINAL DATA/mqc_fastqc_sequence_counts_plot_1.txt') %>%
  filter(!str_detect(Sample,'Atcc')&!str_detect(Sample,'blnk')&!str_detect(Sample,'Zymo')) %>% 
  filter(str_detect(Sample,'bmtagged_1')) %>% 
  mutate(Sample = reformat_ids(Sample) %>% change_ids_to_longitudinal()) %>%
  mutate(depth=`Unique Reads`+`Duplicate Reads`) %>%
  select(Sample,depth)

# Load metadata and convert to phyoseq object~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds")
s = psmelt(ps) %>% select(-OTU,-Abundance,-any_of(c('Kingdom','Phylum','Class','Order',
                                                'Family','Genus','Species'))) %>% unique() %>% 
  mutate(Status = as.factor(Status),
         bristol = as.numeric(bristol),
         Sex = as.factor(Sex),
         ent_boolean = entacapone>0) %>% 
  left_join(cnt) %>% `rownames<-`(.$Sample)

# Taxonomy tables
t = lapply(df,function(x) matrix(data = rownames(x), ncol = 1, dimnames = list(rownames(x),'Species')))

# Filter OTU tables to remove the two failed samples (no bacterial OTUs)
df = df %>% lapply(function(x) x %>% select(all_of(rownames(s))) %>% as.matrix)

# Create phyloseq objects
ps = list()
for(i in 1:4){
  temp = phyloseq(sample_data(s),otu_table(df[[i]],taxa_are_rows = T),tax_table(t[[i]]))
  ps = ps %>% append(temp)
}
names(ps) = names(df)

# Filter enzyme data to only include COG dataset
ps$eggnog = ps$eggnog %>% subset_taxa(str_sub(Species,end=3)=='COG') # COG only

# All done the formatting - everything is now nicely arranged inside a list of phyloseq objects!
rm(cnt,df1,df2,df3,df4,t,temp,i,prep_tsv)
```

*ALDEx2: values will be rounded to nearest integer.*

```{r MetaCyc}
a = Sys.time()
res.s = run_da(ps$pwy,'Status','Species',struc_zero = T, group='Status',prv_cut = 1/3,abun_cut = 0)
res.s.m = run_da(ps$pwy,'Status+Sex+ent_boolean+bristol','Species',struc_zero = T, 
                 group='Status',incldepth = T, stat = 'p_val',prv_cut = 1/3,abun_cut = 0)
species = list(univar = res.s,multivar = res.s.m) %>% lapply(function(x)x$results)

writexl::write_xlsx(species %>% lapply(function(x)x$combined_res),'../Final Outputs/DA Results/DA_PWY_Status_All_Stats_3tools.xlsx')
writexl::write_xlsx(species %>% lapply(function(x)x$significance_summary),'../Final Outputs/DA Results/DA_PWY_Status_All_Summary_3tools.xlsx')
b = Sys.time()
b-a
```

```{r EggNOG}
a = Sys.time()
res.s = run_da(ps$eggnog,'Status','Species',struc_zero = T, group='Status', prv_cut = 0.1, abun_cut = 0)
res.s.m = run_da(ps$eggnog,'Status+Sex+ent_boolean+bristol','Species',struc_zero = T, 
                 group='Status',incldepth = T, stat = 'p_val', prv_cut = 0.1, abun_cut = 0)
species = list(univar = res.s,multivar = res.s.m) %>% lapply(function(x)x$results)

writexl::write_xlsx(species %>% lapply(function(x)x$combined_res),'../Final Outputs/DA Results/DA_EGG_Status_All_Stats_3tools.xlsx')
writexl::write_xlsx(species %>% lapply(function(x)x$significance_summary),'../Final Outputs/DA Results/DA_EGG_Status_All_Summary_3tools.xlsx')
b = Sys.time()
b-a
```

```{r KO}
a = Sys.time()
res.s = run_da(ps$ko,'Status','Species',struc_zero = T, group='Status', prv_cut = 0.1, abun_cut = 0)
res.s.m = run_da(ps$ko,'Status+Sex+ent_boolean+bristol','Species',struc_zero = T, 
                 group='Status',incldepth = T, stat = 'p_val', prv_cut = 0.1, abun_cut = 0)
species = list(univar = res.s,multivar = res.s.m) %>% lapply(function(x)x$results)

writexl::write_xlsx(species %>% lapply(function(x)x$combined_res),'../Final Outputs/DA Results/DA_KO_Status_All_Stats_3tools.xlsx')
writexl::write_xlsx(species %>% lapply(function(x)x$significance_summary),'../Final Outputs/DA Results/DA_KO_Status_All_Summary_3tools.xlsx')
b = Sys.time()
b-a
gc()
```

```{r RXN}
a = Sys.time()
res.s = run_da(ps$rxn,'Status','Species',struc_zero = T, group='Status', prv_cut = 0.1, abun_cut = 0)
res.s.m = run_da(ps$rxn,'Status+Sex+ent_boolean+bristol','Species',struc_zero = T, 
                 group='Status',incldepth = T, stat = 'p_val', prv_cut = 0.1, abun_cut = 0)
species = list(univar = res.s,multivar = res.s.m) %>% lapply(function(x)x$results)

writexl::write_xlsx(species %>% lapply(function(x)x$combined_res),'../Final Outputs/DA Results/DA_RXN_Status_All_Stats_3tools.xlsx')
writexl::write_xlsx(species %>% lapply(function(x)x$significance_summary),'../Final Outputs/DA Results/DA_RXN_Status_All_Summary_3tools.xlsx')
b = Sys.time()
b-a
gc()
```
