---
title: "8. Plot MetaCyc Pathways"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(phyloseq)
library(ggpubr)
library(data.table)
library(tidyverse)

source('../id_functions.R')

prep_tsv = function(df){
  names(df)[1] = 'temp'
  return(df %>% filter(!(temp %in% c('UNMAPPED','UNINTEGRATED','UNGROUPED'))) %>% column_to_rownames('temp') %>%
           select(-contains('Atcc'),-contains('Zymo'),-contains('blnk')) %>%
           `colnames<-`(reformat_ids(colnames(.)) %>% change_ids_to_longitudinal()))
}
# Load RPK Data~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
df = fread('../../../Metagenomics/RPK counts/merged_pathabundance_unstratified.tsv') %>% prep_tsv()

# Load total metagenomic reads~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cnt = read_delim('../../../Metagenomics/ORIGINAL DATA/mqc_fastqc_sequence_counts_plot_1.txt') %>%
  filter(!str_detect(Sample,'Atcc')&!str_detect(Sample,'blnk')&!str_detect(Sample,'Zymo')) %>% 
  filter(str_detect(Sample,'bmtagged_1')) %>% 
  mutate(Sample = reformat_ids(Sample) %>% change_ids_to_longitudinal()) %>%
  mutate(depth=`Unique Reads`+`Duplicate Reads`) %>%
  select(Sample,depth)

# Load metadata and convert to phyoseq object~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds")
s = psmelt(ps) %>% select(-OTU,-Abundance,-any_of(c('Kingdom','Phylum','Class','Order',
                                                'Family','Genus','Species'))) %>% unique() %>% 
  mutate(Status = as.factor(Status),
         bristol = as.numeric(bristol),
         Sex = as.factor(Sex),
         ent_boolean = entacapone>0) %>% 
  left_join(cnt) %>% `rownames<-`(.$Sample)

# Taxonomy tables
t = matrix(data = rownames(df), ncol = 1, dimnames = list(rownames(df),'Species'))

# Filter OTU tables to remove the two failed samples (no bacterial OTUs)
df = df %>% select(all_of(rownames(s))) %>% as.matrix

# Create phyloseq objects
ps = phyloseq(sample_data(s),otu_table(df,taxa_are_rows = T),tax_table(t))
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Plot MetaCyc Pathways}
# Load significant pwys
sig_multi = readxl::read_xlsx('../Final Outputs/DA Results/DA_PWY_Status_All_Summary_3tools.xlsx',sheet='multivar') %>% 
  filter(Variable=='StatusPD',sig_in_at_least_n == T)
sig_uni = readxl::read_xlsx('../Final Outputs/DA Results/DA_PWY_Status_All_Summary_3tools.xlsx',sheet='univar') %>% 
  filter(sig_in_at_least_n == T) %>% filter(Species %in% sig_multi$Species)

# Calculate log 2 fc
lfc = ps %>% microbiome::transform('compositional') %>% psmelt() %>% 
  filter(Species %in% sig_uni$Species) %>% 
  mutate(Species = str_replace(Species, '[&]beta;-D','B-D')) %>% 
  mutate(Species = str_replace(Species, 'superpathway of ', 'superpathway of\n')) %>% 
  mutate(Species = str_replace(Species, ' [(]', '\n(')) %>%
  mutate(Species = str_replace(Species, 'hex-4-', 'hex-4-\n')) %>%
  group_by(Status,Species) %>% summarize(Abundance = mean(Abundance),.groups='keep') %>% ungroup %>% 
  pivot_wider(names_from = Status, values_from = Abundance) %>% 
  mutate(lfc = log2(PD/Ctrl)) %>% arrange(lfc)
lfc$Species = factor(lfc$Species, levels = lfc$Species)

# Plot 1: LFC
p1 = lfc %>% mutate(COL = ifelse(lfc<0,'#01BEC3','#FC746E')) %>% 
  ggplot(aes(lfc,Species,fill = COL)) +
  geom_col(aes(fill = COL)) +
  theme_minimal(base_size = 16) +
  ylab(NULL) + xlab('Log2 FC (PD/Ctrl)') +
  theme(legend.position = 'none')

# Plot 2: Transcripts Per Million
# p2 = ps %>% microbiome::transform('compositional') %>% psmelt() %>% 
#   filter(Species %in% sig_uni$Species) %>% 
#   mutate(Species = str_replace(Species, '[&]beta;-D','beta-D')) %>% 
#   mutate(Species = str_replace(Species, 'superpathway of ', 'superpathway of\n')) %>% 
#   mutate(Species = str_replace(Species, ' [(]', '\n[(]')) %>%
#   mutate(Species = str_replace(Species, 'hex-4-', 'hex-4-\n')) %>%
#   mutate(Abundance = Abundance *1e3,
#          Species = factor(Species, levels = lfc$Species)) %>% 
#   ggplot(aes(Abundance,Species,fill = Status)) +
#   geom_boxplot() +
#   theme_minimal(base_size = 16) +
#   theme(axis.text.y = element_blank()) +
#   xlab('TPM/1000') + ylab(NULL)

# Plot 3: CLR-transformed abundances
p3 = ps %>% microbiome::transform('clr') %>% psmelt() %>% 
  filter(Species %in% sig_uni$Species) %>% 
  mutate(Species = str_replace(Species, '[&]beta;-D','B-D')) %>% 
  mutate(Species = str_replace(Species, 'superpathway of ', 'superpathway of\n')) %>% 
  mutate(Species = str_replace(Species, ' [(]', '\n(')) %>%
  mutate(Species = str_replace(Species, 'hex-4-', 'hex-4-\n')) %>%
  mutate(Species = factor(Species, levels = lfc$Species)) %>% 
  ggplot(aes(Abundance,Species,fill = Status)) +
  geom_boxplot() +
  theme_minimal(base_size = 16) +
  theme(axis.text.y = element_blank()) +
  xlab('CLR') + ylab(NULL)

# p23 = ggarrange(plotlist = list(p2,p3),common.legend = T, ncol = 2, legend = 'right')
# plot.out = ggarrange(plotlist = list(p1,p23),common.legend = F, ncol=2, widths = c(3,2.5))
plot.out = ggarrange(plotlist = list(p1+theme(legend.position='none'),p3),common.legend = F, ncol=2, widths = c(3,1.7))

ggsave('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/PD Functional Analysis/Final Outputs/DA PWY Plots/sig_pwy_3tools.png', height = 7.5, width = 12)
```
