---
title: "9.4. Plot MetaCyc Pathways"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load abundance data}
library(phyloseq)
library(ggpubr)
library(data.table)
library(tidyverse)
library(glue)
library(ggtext)

source('../id_functions.R')

prep_tsv = function(df){
  names(df)[1] = 'temp'
  return(df %>% filter(!(temp %in% c('UNMAPPED','UNINTEGRATED','UNGROUPED'))) %>% column_to_rownames('temp') %>%
           select(-contains('Atcc'),-contains('Zymo'),-contains('blnk')) %>%
           `colnames<-`(reformat_ids(colnames(.)) %>% change_ids_to_longitudinal()))
}
# Load RPK Data~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
df1 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_eggnog_renamed_unstratified.tsv') %>% prep_tsv()
df2 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_ko_renamed_unstratified.tsv') %>% prep_tsv()
df3 = fread('../../../Metagenomics/RPK counts/merged_genefamilies_rxn_renamed_unstratified.tsv') %>% prep_tsv()

# Add to 1 list for high-throughput coding
df = list(eggnog = df1, ko = df2, rxn = df3)

# Load total metagenomic reads~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cnt = read_delim('../../../Metagenomics/ORIGINAL DATA/mqc_fastqc_sequence_counts_plot_1.txt') %>%
  filter(!str_detect(Sample,'Atcc')&!str_detect(Sample,'blnk')&!str_detect(Sample,'Zymo')) %>% 
  filter(str_detect(Sample,'bmtagged_1')) %>% 
  mutate(Sample = reformat_ids(Sample) %>% change_ids_to_longitudinal()) %>%
  mutate(depth=`Unique Reads`+`Duplicate Reads`) %>%
  select(Sample,depth)

# Load metadata and convert to phyoseq object~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds")
s = psmelt(ps) %>% select(-OTU,-Abundance,-any_of(c('Kingdom','Phylum','Class','Order',
                                                'Family','Genus','Species'))) %>% unique() %>% 
  mutate(Status = as.factor(Status),
         bristol = as.numeric(bristol),
         Sex = as.factor(Sex),
         ent_boolean = entacapone>0) %>% 
  left_join(cnt) %>% `rownames<-`(.$Sample)

# Taxonomy tables
t = lapply(df,function(x) matrix(data = rownames(x), ncol = 1, dimnames = list(rownames(x),'Species')))

# Filter OTU tables to remove the two failed samples (no bacterial OTUs)
df = df %>% lapply(function(x) x %>% select(all_of(rownames(s))) %>% as.matrix)

# Create phyloseq objects
ps = list()
for(i in 1:3){
  temp = phyloseq(sample_data(s),otu_table(df[[i]],taxa_are_rows = T),tax_table(t[[i]]))
  ps = ps %>% append(temp)
}
names(ps) = names(df)

# Filter enzyme data to only include COG dataset
ps$eggnog = ps$eggnog %>% subset_taxa(str_sub(Species,end=3)=='COG') # COG only

# All done the formatting - everything is now nicely arranged inside a list of phyloseq objects!
rm(cnt,df1,df2,df3,t,temp,i,prep_tsv)
```

#\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~\~

```{r}
cog_key = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/PD Functional Analysis/Reference Files/COG_key.csv') %>% 
  select(Cat,desc) %>% rename(Test=Cat) %>% mutate(desc = str_to_title(desc)) %>% 
  unique
cog = readxl::read_xlsx('../Final Outputs/Enrichment Analyses/COG_3Tools_COGonly.xlsx',sheet='All Results p<0.1') %>% 
  filter(qval<0.05) %>% 
  mutate(Dataset = 'COG') %>% left_join(cog_key) %>% 
  mutate(Test = desc)

ko = readxl::read_xlsx('../Final Outputs/Enrichment Analyses/KO_3Tools_allsig.xlsx',sheet='KEGG') %>% 
  filter(qval<0.05) %>% 
  mutate(Dataset = 'KEGG',
         Level = ifelse(!is.na(L4),'L4',ifelse(!is.na(L3),'L3','L2')))

ec_key = read.csv('C:/Users/armetcal/OneDrive - UBC/Grad School/Data/HUMAN COHORT/PD Functional Analysis/Reference Files/EC_key.csv')
ec = readxl::read_xlsx('../Final Outputs/Enrichment Analyses/RXN_3Tools.xlsx',sheet='All Results p<0.1') %>% 
  filter(qval<0.05) %>% 
  mutate(Dataset = 'EC')

enrich = cog %>% full_join(ko) %>% full_join(ec) %>% 
  mutate(Enrichment = log2(pct_pwy/pct_total),
         n_pwy = pwy_sig + pwy_ns) %>% 
  select(Dir,Level,Test,Dataset,L1,L2,L3,L4,Enrichment,n_pwy) %>% 
  mutate(Test = ifelse(Dataset=='KEGG' & !is.na(L4),L4,
                       ifelse(Dataset=='KEGG' & !is.na(L3),L3,
                              ifelse(Dataset=='KEGG' & !is.na(L2),L2,Test)))) %>% 
  arrange(Level) %>% arrange(L2) %>% arrange(Dataset) %>% 
  mutate(Enrichment = ifelse(Dir == 'Negative',Enrichment*-1,Enrichment))

```

```{r Formatting the table}
enrich2 = enrich %>% 
  mutate(cat = ifelse(Dataset=='COG','Category',NA)) %>% 
  mutate(cat = ifelse(Dataset=='EC' & str_sub(Test,end=1)=='1',
                      ec_key$class.name[ec_key$Class=='1'] %>% unique,cat)) %>% 
  mutate(cat = ifelse(Dataset=='EC' & str_sub(Test,end=1)=='2',
                      ec_key$class.name[ec_key$Class=='2'] %>% unique,cat)) %>% 
  mutate(cat = ifelse(Dataset=='EC' & str_sub(Test,end=1)=='3',
                      ec_key$class.name[ec_key$Class=='3'] %>% unique,cat)) %>% 
  mutate(cat = ifelse(Dataset=='KEGG' & L2 == '09120 Genetic Information Processing',
                      'Gen. Info Processing',cat)) %>% 
  mutate(cat = ifelse(Dataset=='KEGG' & L2 == '09130 Environmental Information Processing',
                      'Env. Info Processing',cat)) %>% 
  mutate(cat = ifelse(Dataset=='KEGG' & L2 == '09140 Cellular Processes',
                      'Cellular Processes',cat)) %>% 
  mutate(cat = ifelse(Dataset=='KEGG' & L2 == '09180 Brite Hierarchies',
                      'Brite Hierarchies',cat))

enrich2 = enrich2 %>% 
  mutate(Test = ifelse(Test=='3.1.4','3.1.4 Phosphoric Diester Hydrolases',
                ifelse(Test=='2.7.4','2.7.4 Phosphotransferases, Phosphate Acceptor',
                ifelse(Test=='1','1 Oxidoreductases',
                ifelse(Test=='1.1','1.1 Acting on CH-OH',
                ifelse(Test=='1.1.1','1.1.1 NAD(+)/NADP(+) Acceptor',
                ifelse(Test=='09183 Protein families: signaling and cellular processes','09183 Signaling/Cellular Processes',
                ifelse(Test=='09130 Environmental Information Processing','09130 Env. Info Processing',Test)))))))) %>% 
  mutate(bold = ifelse(Test %in% c('1 Oxidoreductases','2.7.4 Phosphotransferases, Phosphate Acceptor','3.1.4 Phosphoric Diester Hydrolases','09122 Translation','09130 Env. Info Processing','09142 Cell motility','09183 Signaling/Cellular Processes','Signal Transduction Mechanisms','Cell Motility','Carbohydrate Transport And Metabolism'),glue("<b>{Test}</b>"),Test))

order = c(rev(enrich2$bold[enrich2$Dataset=='COG']),
            rev(enrich2$bold[enrich2$cat=='Oxidoreductases']),
            '<b>2.7.4 Phosphotransferases, Phosphate Acceptor</b>',
            '<b>3.1.4 Phosphoric Diester Hydrolases</b>',
            rev(enrich2$bold[enrich2$Dataset=='KEGG']))
enrich2$bold = factor(enrich2$bold, levels = order)

enrich2$cat = factor(enrich2$cat, levels = c('Category','Oxidoreductases','Transferases','Hydrolases',sort(enrich2$cat[enrich2$Dataset=='KEGG'] %>% unique)))
```

```{r}
enrich3 = enrich2 %>% 
  arrange(bold) %>%
  group_by(cat) %>% 
  mutate(prefix = Test[n()]) %>% ungroup %>% 
  # mutate(prefix = ifelse(Dataset!='KEGG','',paste0(str_sub(prefix,end=5),' \u2192 '))) %>% 
  # mutate(prefix = ifelse(Dataset!='KEGG','','\u2192 ')) %>% 
  mutate(prefix = '\u2192 ') %>% 
  mutate(bold2 = as.character(bold)) %>% 
  mutate(bold2 = ifelse(!is.na(prefix) & str_sub(bold2,end=3) !='<b>',
                        paste0(prefix,bold2),bold2)) %>% 
  mutate(bold2 = ifelse(Test %in% c('02010 ABC transporters','1.1.1 NAD(+)/NADP(+) Acceptor'),paste0(prefix,bold2),bold2)) %>% 
  arrange(bold)

enrich3$bold2 = factor(enrich3$bold2, levels = enrich3$bold2)
```

```{r}
library(ggh4x)
enrich3 %>%
  ggplot(aes(Enrichment,bold2,fill=Dir)) +
  geom_col() +
  theme_classic(base_size=16) +
  facet_nested(Dataset + cat ~ .,scales = 'free',space = 'free') +
  ylab(NULL) + xlab('Log 2 Fold Enrichment (PD/Ctrl)') +
  theme(strip.text.y.right = element_text(angle = 0)) +
  theme(axis.text.y=element_markdown()) +
  theme(legend.position = 'none') +
  theme(panel.spacing=unit(1,"lines"))

ggsave('../Final Outputs/Enrichment Analyses/9.4. Plot Enriched Pathways - 3 Tools.jpeg',height=8, width = 12)
```
