---
title: "10. Determine Key Species Per Pathway"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(phyloseq)
library(data.table)
library(ggpubr)
library(tidyverse)
library(gt)

source('../id_functions.R')

prep_tsv = function(df){
  names(df)[1] = 'temp'
  return(df %>% filter(!str_detect(temp,'UNMAPPED')) %>%
           filter(!str_detect(temp,'UNINTEGRATED')) %>% 
           filter(!str_detect(temp,'UNGROUPED')) %>% 
           column_to_rownames('temp') %>%
           select(-contains('Atcc'),-contains('Zymo'),-contains('blnk')) %>%
           `colnames<-`(reformat_ids(colnames(.)) %>% change_ids_to_longitudinal()))
}
# Load RPK Data~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
df = fread('../../../Metagenomics/RPK counts/merged_pathabundance_stratified.tsv') %>% 
  select(all_of(colnames(.)[-which(duplicated(colnames(.)))])) %>% prep_tsv()

# Load total metagenomic reads~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cnt = read_delim('../../../Metagenomics/ORIGINAL DATA/mqc_fastqc_sequence_counts_plot_1.txt') %>%
  filter(!str_detect(Sample,'Atcc')&!str_detect(Sample,'blnk')&!str_detect(Sample,'Zymo')) %>% 
  filter(str_detect(Sample,'bmtagged_1')) %>% 
  mutate(Sample = reformat_ids(Sample) %>% change_ids_to_longitudinal()) %>%
  mutate(depth=`Unique Reads`+`Duplicate Reads`) %>%
  select(Sample,depth)

# Load metadata and convert to phyoseq object~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds")
s = psmelt(ps) %>% select(-OTU,-Abundance,-any_of(c('Kingdom','Phylum','Class','Order',
                                                'Family','Genus','Species'))) %>% unique() %>% 
  mutate(Status = as.factor(Status),
         bristol = as.numeric(bristol),
         Sex = as.factor(Sex),
         ent_boolean = entacapone>0) %>% 
  left_join(cnt) %>% `rownames<-`(.$Sample)

# Taxonomy tables
t = data.frame(Species = rownames(df), row.names = rownames(df)) %>% 
  separate(Species,into=c('feature','taxon'),sep='[|]') %>% as.matrix

# Filter OTU tables to remove the two failed samples (no bacterial OTUs)
df = df %>% select(all_of(rownames(s))) %>% as.matrix

# Create phyloseq objects
ps = phyloseq(sample_data(s),otu_table(df,taxa_are_rows = T),tax_table(t))
```

```{r Select pathways of interest}
sig_multi = readxl::read_xlsx('../Final Outputs/DA Results/DA_PWY_Status_All_Summary_3tools.xlsx',sheet='multivar') %>% 
  filter(Variable=='StatusPD',sig_in_at_least_n == T)
sig_uni = readxl::read_xlsx('../Final Outputs/DA Results/DA_PWY_Status_All_Summary_3tools.xlsx',sheet='univar') %>% 
  filter(sig_in_at_least_n == T) %>% filter(Species %in% sig_multi$Species)
sig_uni.coef = readxl::read_xlsx('../Final Outputs/DA Results/DA_PWY_Status_All_Stats_3tools.xlsx',sheet='univar') %>% 
  filter(Species %in% sig_multi$Species) %>% select(Species,association_dir) %>% unique

df2 = ps %>% microbiome::transform('clr') %>% 
  subset_taxa(feature %in% sig_uni$Species) %>% psmelt()
df3 = ps %>% microbiome::transform('compositional') %>% 
  subset_taxa(feature %in% sig_uni$Species) %>% psmelt()
```

# Just a check to ensure that the log fold changes are very similar to the unstratified version

Everything looks decent! Some differences, nothing too crazy. 

```{r Log 2 FC with compositional abundance}
mean = df3 %>% 
  group_by(feature,Sample,Status) %>% 
  summarize(Abundance = sum(Abundance),.groups = 'keep') %>% ungroup() %>% 
  group_by(feature,Status) %>% 
  filter(!is.nan(Abundance)) %>% 
  summarize(Abundance = mean(Abundance), .groups = 'keep') %>% ungroup() %>% 
  pivot_wider(names_from = Status, values_from = Abundance) %>% 
  mutate(Diff = PD-Ctrl, log2FC = log2(PD/Ctrl))
```

# Differential Abundance with CLR

```{r}
uni = df2 %>% group_by(feature,taxon) %>% 
  group_modify(~summary(glm(Abundance~Status, data = .))$coefficients %>% 
                 as.matrix() %>% as.data.frame() %>% rownames_to_column('var')) %>% 
  ungroup() %>% rename(pval = `Pr(>|t|)`) %>% 
  filter(!str_detect(var,'Intercept')) %>% 
  group_by(var) %>% 
  mutate(qval = p.adjust(pval,method = 'fdr'), .after = pval) %>% 
  ungroup() %>% 
  mutate(Type = 'Univariate')

mult = df2 %>% group_by(feature,taxon) %>% 
  group_modify(~summary(glm(Abundance~Status+Sex+bristol+ent_boolean+depth, data = .))$coefficients %>% 
                 as.matrix() %>% as.data.frame() %>% rownames_to_column('var')) %>% 
  ungroup() %>% rename(pval = `Pr(>|t|)`) %>% 
  filter(!str_detect(var,'Intercept')) %>% 
  group_by(var) %>% # correct p values per variable
  mutate(qval = p.adjust(pval,method = 'fdr'), .after = pval) %>% 
  ungroup() %>% 
  mutate(Type = 'Multivariate')

unisig = uni %>% filter(qval<0.05) %>% left_join(sig_uni.coef %>% rename(feature=Species))
multisig = mult %>% filter(qval<0.05) %>% left_join(sig_uni.coef %>% rename(feature=Species))

# Filter to only include stuff that is different in uni and multivar results
multisig_prefilt = mult %>% filter(var=='StatusPD', paste0(feature,taxon) %in% paste0(unisig$feature,unisig$taxon)) %>% 
  mutate(qval = p.adjust(pval, method = 'fdr')) %>% filter(qval<0.05) %>% arrange(pval) %>% arrange(qval) %>% left_join(sig_uni.coef %>% rename(feature=Species))

# Filter to only include results where the taxon-specific direction matches the overall direction of the pathway (otherwise it's not driving effect)
unisig2 = unisig %>% filter((association_dir=='Positive')==(Estimate>0))
multisig_prefilt2 = multisig_prefilt %>% filter((association_dir=='Positive')==(Estimate>0))
```
```{r Save raw statistics}
writexl::write_xlsx(list(Univariate = uni, Multivariate = mult),
                    '../Final Outputs/DA Results/10. Species-Stratified Pathway Changes.xlsx')
```


```{r Make a nice table - formatting}
stats = rbind(unisig2,multisig_prefilt2) %>% 
  # left_join(mean) %>% 
  rename(Pathway = feature,Species=taxon,Variable = var) %>% 
  pivot_wider(values_from = c(Estimate:qval), names_from = Type) %>% 
  mutate(Pathway = str_replace(Pathway, '[&]beta;-D','B-D')) %>% 
  mutate(Pathway = str_replace(Pathway, 'superpathway of ', 'superpathway of\n')) %>% 
  mutate(Pathway = str_replace(Pathway, ' [(]', '\n[(]')) %>%
  mutate(Pathway = str_replace(Pathway, 'hex-4-', 'hex-4-\n'))
```

```{r make pretty with gt}
gt_tbl = stats %>% select(Pathway, everything()) %>%
  mutate(Species = str_remove(Species,'.*[.]') %>% str_remove_all('s__') %>% str_replace_all('_',' ') %>% str_replace('uncl','Uncl')) %>% 
  mutate_if(is.numeric,signif,3) %>% 
  mutate_at(vars(contains('pval'),contains('qval')),scales::scientific,3) %>% 
  arrange(pval_Univariate) %>% arrange(Species) %>% arrange(association_dir) %>% 
  gt() %>% 
  tab_row_group(label = md('**Pwy Elevated in PD**'),rows = (association_dir=='Positive')) %>% 
  tab_row_group(label = md('**Pwy Reduced in PD**'),rows = (association_dir=='Negative')) %>% 
  cols_hide(c('association_dir','Variable')) %>% 
  tab_spanner('Estimate',columns = contains('Estimate')) %>% 
  tab_spanner('Std Error',columns = contains('Error')) %>% 
  tab_spanner('t Value',columns = contains('t value')) %>% 
  tab_spanner('P Value',columns = contains('pval')) %>% 
  tab_spanner('Q Value',columns = contains('qval')) %>% 
  cols_label(
    `Estimate_Univariate` = "Univar",
    `Estimate_Multivariate` = "Multivar",
    `Std. Error_Univariate` = 'Univar',
    `Std. Error_Multivariate` = 'Multivar',
    `t value_Univariate` = "Univar",
    `t value_Multivariate` = 'Multivar',
    `pval_Univariate` = "Univar",
    `pval_Multivariate` = 'Multivar',
    `qval_Univariate` = "Univar",
    `qval_Multivariate` = 'Multivar'
  ) %>%
  cols_align('center',-c(Pathway,Species)) %>% 
  tab_options(column_labels.font.weight = 'bold')
gt_tbl
# Manually save the html as a PDF
gtsave(gt_tbl,filename = "../Final Outputs/DA Results/Key Species per Pathway - 3 Tools.html")
```

# EXTRA ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
```{r Test whether amount of pathway within F. prausnitzii is different between groups}
# # I.e. if we only consider people with detectable F. prausnitzii, is the difference still present?
# 
# # Combinations of interest
# comb = stats %>% filter(Species != 'Unclassified')
# 
# # Load species abundance data, normalize according to Maaslin2 parameters
# ps = readRDS("../Reference Files/phyloseq_object_leafyadj_absolute_counts.rds") %>% 
#   microbiome::transform('compositional') %>%
#   subset_taxa(Species == 's__Faecalibacterium_prausnitzii') %>% psmelt
# 
# # Make combined dataset
# df4 = df3 %>% filter(feature %in% comb$Pathway, str_detect(taxon, 'prausnitzii')) %>% 
#   rename(ab_path = Abundance) %>% 
#   left_join(ps %>% select(Sample, Abundance) %>% rename(ab_tax = Abundance)) %>% 
#   filter(!is.na(Status))
# 
# # Taxon vs taxon-specific pathway
# df4 %>% ggplot(aes(ab_path,ab_tax, col = Status, fill = Status)) +
#   geom_point() +
#   geom_smooth(method = 'lm') +
#   ggpubr::stat_cor(method = 'spearman') +
#   theme_classic(base_size = 20) +
#   ylab('F. prausnitzii') + xlab('Pathway') +
#   facet_wrap('feature')
# 
# # Taxon-normalized pathway. Any zeros (no prausnitzii) are removed.
# df5 = df4 %>% filter(!(ab_tax==0 & ab_path==0)) %>% 
#   mutate(Abundance = ab_path/ab_tax)
# table(is.nan(df5$Abundance)) # ensure no calculation errors
# 
# df5 %>% mutate(OTU=str_wrap(OTU,30)) %>% 
#   ggplot(aes(OTU,Abundance, fill = Status)) +
#   geom_boxplot(outlier.shape = NA) +
#   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2)) +
#   ggpubr::stat_compare_means(aes(group = Status), size = 5, label = 'p.signif') +
#   theme_classic(base_size=10) +
#   ylab('Pathway/F. prausnitzii') + xlab(NULL) +
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

